Special handling for classes catching LinkageError

When checking jib-maven-plugin (which uses maven-core), the enforcer rule outputs the following message.

```
[ERROR] Linkage Checker rule found 6 reachable errors. Linkage error report:
Class com.google.inject.servlet.ServletModuleTargetVisitor is not found;
  referenced by 1 class file
    org.eclipse.sisu.inject.Implementations (org.eclipse.sisu.inject-0.3.3.jar)
Class com.google.inject.servlet.InstanceFilterBinding is not found;
  referenced by 1 class file
    org.eclipse.sisu.inject.Implementations (org.eclipse.sisu.inject-0.3.3.jar)
Class com.google.inject.servlet.InstanceServletBinding is not found;
  referenced by 1 class file
    org.eclipse.sisu.inject.Implementations (org.eclipse.sisu.inject-0.3.3.jar)
Class com.google.inject.servlet.LinkedFilterBinding is not found;
  referenced by 1 class file
    org.eclipse.sisu.inject.Implementations (org.eclipse.sisu.inject-0.3.3.jar)
Class com.google.inject.servlet.LinkedServletBinding is not found;
  referenced by 1 class file
    org.eclipse.sisu.inject.Implementations (org.eclipse.sisu.inject-0.3.3.jar)
Class javax.annotation.Priority is not found;
  referenced by 1 class file
    org.eclipse.sisu.inject.Implementations (org.eclipse.sisu.inject-0.3.3.jar)
```

The `Implementations` class checks the existence of several classes by catching `LinkageError`.

![image](https://user-images.githubusercontent.com/28604/61892145-31506780-aed9-11e9-8673-4f62a077ce9d.png)


We can handle such case in the same manner as we did #402 for logger classes.

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java
@@ -478,11 +478,14 @@ class ClassDumper {
     return constantPoolIndicesForTarget.build();
   }
 
+  private static final ImmutableSet<String> ERRORS_CAUGHT_IN_SOURCE =
+      ImmutableSet.of(NoClassDefFoundError.class.getName(), LinkageError.class.getName());
+
   /**
-   * Returns true if {@code sourceClassName} has a method that has an exception
-   * handler for {@link NoClassDefFoundError}.
+   * Returns true if {@code sourceClassName} has a method that has an exception handler for {@link
+   * NoClassDefFoundError} or {@link LinkageError}.
    */
-  boolean catchesNoClassDefFoundError(String sourceClassName) {
+  boolean catchesLinkageError(String sourceClassName) {
     try {
       JavaClass sourceJavaClass = loadJavaClass(sourceClassName);
       ClassGen classGen = new ClassGen(sourceJavaClass);
@@ -496,11 +499,19 @@ class ClassDumper {
             continue;
           }
           String caughtClassName = catchType.getClassName();
-          if (NoClassDefFoundError.class.getName().equals(caughtClassName)) {
-            // NoClassDefFoundError is caught in the source class
+          if (ERRORS_CAUGHT_IN_SOURCE.contains(caughtClassName)) {
+            // The source class catches an error and thus will not cause a runtime error
             return true;
           }
         }
+      }
+
+      String outerClassName = outerClassName(sourceJavaClass);
+      if (outerClassName != null) {
+        return catchesLinkageError(outerClassName);
+      } else {
+        // The source class does not have a method that catches NoClassDefFoundError
+        return false;
       }
     } catch (ClassNotFoundException ex) {
       // Because the reference in the argument was extracted from the source class file,
@@ -508,9 +519,37 @@ class ClassDumper {
       throw new ClassFormatException(
           "The source class in the reference is no longer available in the class path", ex);
     }
+  }
 
-    // The source class does not have a method that catches NoClassDefFoundError
-    return false;
+  private static String outerClassName(JavaClass sourceJavaClass) {
+    ConstantPool constantPool = sourceJavaClass.getConstantPool();
+    if (sourceJavaClass.isNested()) {
+      for (Attribute attribute : sourceJavaClass.getAttributes()) {
+        if (attribute instanceof InnerClasses) {
+          InnerClasses innerClasses = (InnerClasses) attribute;
+          for (InnerClass innerClass : innerClasses.getInnerClasses()) {
+            // Some auto-generated classes may not have correct inner class entries,
+            // for example com.google.inject.internal.cglib.core.$DebuggingClassWriter.java
+            if (innerClass.getInnerClassIndex() <= 0 || innerClass.getOuterClassIndex() <= 0) {
+              continue;
+            }
+            // Class names in constant pool has '/' as separator
+            String innerClassName =
+                constantPool
+                    .getConstantString(innerClass.getInnerClassIndex(), Const.CONSTANT_Class)
+                    .replaceAll("/", ".");
+            String outerClassName =
+                constantPool
+                    .getConstantString(innerClass.getOuterClassIndex(), Const.CONSTANT_Class)
+                    .replaceAll("/", ".");
+            if (innerClassName.equals(sourceJavaClass.getClassName())) {
+              return outerClassName;
+            }
+          }
+        }
+      }
+    }
+    return null;
   }
 
   /**

--- a/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java
+++ b/dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageChecker.java
@@ -217,7 +217,7 @@ public class LinkageChecker {
       return Optional.of(
           new SymbolProblem(symbol, ErrorType.SYMBOL_NOT_FOUND, containingClassFile));
     } catch (ClassNotFoundException ex) {
-      if (classDumper.catchesNoClassDefFoundError(sourceClassName)) {
+      if (classDumper.catchesLinkageError(sourceClassName)) {
         return Optional.empty();
       }
       ClassSymbol classSymbol = new ClassSymbol(symbol.getClassName());
@@ -263,7 +263,7 @@ public class LinkageChecker {
       return Optional.of(
           new SymbolProblem(symbol, ErrorType.SYMBOL_NOT_FOUND, containingClassFile));
     } catch (ClassNotFoundException ex) {
-      if (classDumper.catchesNoClassDefFoundError(sourceClassName)) {
+      if (classDumper.catchesLinkageError(sourceClassName)) {
         return Optional.empty();
       }
       ClassSymbol classSymbol = new ClassSymbol(symbol.getClassName());
@@ -345,7 +345,7 @@ public class LinkageChecker {
       return Optional.empty();
     } catch (ClassNotFoundException ex) {
       if (classDumper.isUnusedClassSymbolReference(sourceClassName, symbol)
-          || classDumper.catchesNoClassDefFoundError(sourceClassName)) {
+          || classDumper.catchesLinkageError(sourceClassName)) {
         // The class reference is unused in the source
         return Optional.empty();
       }

--- a/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
+++ b/dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerTest.java
@@ -741,6 +741,26 @@ public class LinkageCheckerTest {
   }
 
   @Test
+  public void testFindSymbolProblems_catchesLinkageError() throws RepositoryException, IOException {
+    // org.eclipse.sisu.inject.Implementations catches LinkageError to detect the availability of
+    // implementation for dependency injection. The tool should not show errors for such classes.
+    List<Path> paths =
+        ClassPathBuilder.artifactsToClasspath(
+            ImmutableList.of(
+                new DefaultArtifact("org.eclipse.sisu:org.eclipse.sisu.inject:0.3.3")));
+
+    LinkageChecker linkageChecker = LinkageChecker.create(paths, paths);
+
+    ImmutableSetMultimap<ClassFile, SymbolProblem> problems =
+        linkageChecker.findSymbolProblems().inverse();
+
+    Path sisuJar = paths.get(0);
+    Truth.assertThat(
+            problems.get(new ClassFile(sisuJar, "org.eclipse.sisu.inject.Implementations")))
+        .isEmpty();
+  }
+
+  @Test
   public void testFindSymbolProblems_doesNotCatchNoClassDefFoundError() throws IOException {
     // Checking Firestore jar file without its dependency should have linkage errors
     // Note that FirestoreGrpc.java does not have catch clause of NoClassDefFoundError

